crates/bevy_ecs/src/system/query.rs

```diff
@@ -1521,6 +1521,58 @@ impl<'w, 's, D: QueryData, F: QueryFilter> Query<'w, 's, D, F> {
     pub fn as_query_lens(&mut self) -> QueryLens<'_, D> {
         self.transmute_lens()
     }
+
+    /// Returns a QueryLens that can be used to get a query with the combined fetch.
+    ///
+    /// For example, this can take a `Query<&A>` and a `Queryy<&B>` and return a `Query<&A, &B>`.
+    /// The returned query will only return items with both `A` and `B`. Note that since filter
+    /// are dropped, non-archetypal filters like `Added` and `Changed` will no be respected.
+    /// To maintain or change filter terms see `Self::join_filtered`.
+    ///
+    /// ## Panics
+    ///
+    /// This will panic if `NewD` is not a subset of the union of the original fetch `Q` and `OtherD`.
+    ///
+    /// ## Allowed Transmutes
+    ///
+    /// Like `transmute_lens` the query terms can be changed with some restrictions.
+    /// See [`Self::transmute_lens`] for more details.
+    pub fn join<OtherD: QueryData, NewD: QueryData>(
+        &mut self,
+        other: &Query<OtherD>,
+    ) -> QueryLens<'_, NewD> {
+        self.join_filtered(other)
+    }
+
+    /// Equivalent to [`Self::join`] but also includes a `QueryFilter` type.
+    ///
+    /// Note that the lens with iterate a subset of the original queries tables
+    /// and arcchetypes. This means that additional archetypal query terms like
+    /// `With` and `Without` will not necessarily be respected and non-archetypal
+    /// terms like `Added` and `Changed` will only be respected if they are in
+    /// the type signature.
+    pub fn join_filtered<
+        OtherD: QueryData,
+        OtherF: QueryFilter,
+        NewD: QueryData,
+        NewF: QueryFilter,
+    >(
+        &mut self,
+        other: &Query<OtherD, OtherF>,
+    ) -> QueryLens<'_, NewD, NewF> {
+        // SAFETY: There are no other active borrows of data from world
```

Comment:
This safety comment seems inadequate to me. This is a method on `Query`, which is almost always used while other systems and system params are active.

Reply:
My gut reaction is that this is probably sound, but the actual reason *why* is because `Query` cannot coexist while archetypes are getting shuffled.

However, this makes me very worried about change detection filters. What happens if we ask to filter by Changed<T>, and T isn't mutably borrowed by the original queries? We have no lock, but need to access information that could be simultaneously changed by other system param or systems.

Reply:
Ah, yeah I copied this line from transmute_filtered. I'll try to improve it. I'll also try adding a test to see if we get a panic adding a Changed<T> for one we don't already have read access to.

Reply:
>  What happens if we ask to filter by Changed, and T isn't mutably borrowed by the original queries?

Changed takes read access, so this is fine.

Added a couple tests and cleaned up the safety comment. Fixed some issues along the way